### ISSUE TITLE
runMSE errors when calculating F from TAC

### ISSUE THREAD

**Issue by quang-huynh** [Blue-Matter/MSEtool#36]
I've found two error messages from `runMSE` when running the SAMtool examples under \donttest (see https://github.com/Blue-Matter/SAMtool/runs/1724495484?check_suite_focus=true). Unfortunately they're not replicable on my laptop.

```
The following error occured when running the forward projections: Error 
  in expC[TACusedE > availB] <- availB[TACusedE > availB] * 0.99: NAs are 
  not allowed in subscripted assignments 
```
```
The following error occured when running the forward projections: Error 
  in if (ft <= 0.001) return(tiny): missing value where TRUE/FALSE needed 
```

My hunch is that is an issue with calculating F from a TAC recommendation when the biomass is very low. Non-reproducibility indicates a memory leaks in TMB that I have not been able to find.

**Comment by sebpardo** [comment]
Hi @quang-huynh, I was able to replicate this error in my computer:

```
> library(DLMtool)
Loading required package: MSEtool
Loading required package: snowfall
Loading required package: snow
> myMSE <- runMSE()
Argument `MPs=NA`, using all example MPs in `MSEtool` 

Searching for objects of class MP in package: MSEtool 

Checking MPs 

Loading operating model 

Optimizing for user-specified movement 

Optimizing for user-specified depletion in last historical year 

Calculating historical stock and fishing dynamics 

Calculating MSY reference points for each year 

Calculating B-low reference points 

Calculating reference yield - best fixed F strategy 

Simulating observed data 

Running forward projections 

1 / 5 Running MSE for curEref 

..................................................
2 / 5 Running MSE for FMSYref 

..............The following error occured when running the forward projections: Error 
in if (ft <= 0.001) return(tiny): missing value where TRUE/FALSE needed  

Returning the historical simulations (class `Hist`). To avoid re-running 
spool up, the forward projections can be run with `runMSE(Hist, MPs, ...)` 
```
The error occurs repeatedly and when running MSEs for different MPs, but eventually if ran enough times, there code will run with no errors. In my limited tests, it runs successfully 4 out of 20 trials:

```
> runMSE_class <- replicate(20, class(runMSE()))
> table(runMSE_class)
runMSE_class
Hist  MSE 
  16    4 
```

If you're interested, here is my session info:
```
> sessionInfo()
R version 4.0.3 (2020-10-10)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/atlas/libblas.so.3.10.3
LAPACK: /usr/lib/x86_64-linux-gnu/atlas/liblapack.so.3.10.3

locale:
 [1] LC_CTYPE=en_CA.UTF-8       LC_NUMERIC=C               LC_TIME=en_CA.UTF-8       
 [4] LC_COLLATE=en_CA.UTF-8     LC_MONETARY=en_CA.UTF-8    LC_MESSAGES=en_CA.UTF-8   
 [7] LC_PAPER=en_CA.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_CA.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] DLMtool_6.0.0     MSEtool_3.0.0     snowfall_1.84-6.1 snow_0.4-3       

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.6       rstudioapi_0.13  magrittr_2.0.1   munsell_0.5.0   
 [5] tidyselect_1.1.0 colorspace_2.0-0 R6_2.5.0         rlang_0.4.10    
 [9] dplyr_1.0.3      tools_4.0.3      parallel_4.0.3   grid_4.0.3      
[13] gtable_0.3.0     DBI_1.1.1        ellipsis_0.3.1   abind_1.4-5     
[17] assertthat_0.2.1 tibble_3.0.5     lifecycle_0.2.0  crayon_1.3.4    
[21] purrr_0.3.4      ggplot2_3.3.3    vctrs_0.3.6      glue_1.4.2      
[25] compiler_4.0.3   pillar_1.4.7     scales_1.1.1     generics_0.1.0  
[29] pkgconfig_2.0.3 
```
If there is any way I can help you debug this issue please let me know, I'd be more than happy to help.

**Comment by sebpardo** [comment]
I should also mention that I'm getting both errors when running the 20 replicates, and not just the missing value one.

**Comment by AdrianHordyk** [comment]
Thanks @sebpardo - that's helpful to confirm the problem occurs more often on Linux systems.

This should be fixed in the latest patch in the dev version MSEtool v3.0.9004

It was caused by a memory leak in the C++ pop dynamics code which seems to occur much more commonly on Linux systems.

`devtools::install_github('blue-matter/MSEtool')` should fix it.

@quang-huynh I'll push a patch to CRAN soon which should fix the error in the projections when running `SAMtool` MPs, though I had a similar and I think unrelated problem when applying some `SAMtool` MPs to Data with large number of simulations. If the problem is still there, it will likely be much easier to debug on a Linux system

Closing issue, but please re-open if the problem still occurs with v3.0.9004

**Comment by sebpardo** [comment]
Thanks @AdrianHordyk, just letting you know that the issue is indeed fixed after installing v3.0.9004 in my Linux machine.
